drivers/usb/udc/udc_ambiq.c

```diff
@@ -0,0 +1,925 @@
+/*
+ * Copyright 2024 Ambiq Micro Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <soc.h>
+#include "am_mcu_apollo.h"
+#include <string.h>
+#include <zephyr/drivers/clock_control/clock_control_ambiq.h>
+#include <zephyr/sys/util.h>
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/usb/udc.h>
+#include <zephyr/sys/util.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/cache.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/irq.h>
+#include "udc_common.h"
+#include <usb_dwc2_hw.h>
+
+LOG_MODULE_REGISTER(udc_ambiq, CONFIG_UDC_DRIVER_LOG_LEVEL);
+
+enum udc_ambiq_event_type {
+	/* SETUP packet received at Control Endpoint */
+	UDC_AMBIQ_EVT_HAL_SETUP,
+	/* OUT transaction completed */
+	UDC_AMBIQ_EVT_HAL_OUT_CMP,
+	/* IN transaction completed */
+	UDC_AMBIQ_EVT_HAL_IN_CMP,
+	/* Xfer request received via udc_ambiq_ep_enqueue API */
+	UDC_AMBIQ_EVT_XFER,
+};
+
+struct udc_ambiq_event {
+	const struct device *dev;
+	enum udc_ambiq_event_type type;
+	uint8_t ep;
+};
+
+K_MSGQ_DEFINE(drv_msgq, sizeof(struct udc_ambiq_event), CONFIG_UDC_AMBIQ_MAX_QMESSAGES,
+	      sizeof(void *));
+
+/* USB device controller access from devicetree */
+#define DT_DRV_COMPAT ambiq_usb
+
+#define EP0_MPS   64U
+#define EP_FS_MPS 64U
+#define EP_HS_MPS 512U
+
+struct udc_ambiq_data {
+	struct k_thread thread_data;
+	void *usb_handle;
+	am_hal_usb_dev_speed_e usb_speed;
+	uint8_t setup[8];
+};
+
+struct udc_ambiq_config {
+	uint32_t num_endpoints;
+	int speed_idx;
+	struct udc_ep_config *ep_cfg_in;
+	struct udc_ep_config *ep_cfg_out;
+	void (*make_thread)(const struct device *dev);
+	void (*irq_enable_func)(const struct device *dev);
+	void (*irq_disable_func)(const struct device *dev);
+};
+
+const static struct device *udc_ambiq_dev;
```

Comment:
How does it work if you have multiple instances?

Reply:
There is only 1 USB device controller in Ambiq chips. 
I believe that there will not be multiple instances for UDC for those chips.

Reply:
Looks like bad HAD driver design. But variables below can be moved to udc_ambiq_data.

Reply:
Yes. You are right, that was indeed a bad driver design!
I have removed the dependency on `udc_ambiq_dev` of the functions in this driver, and moved all the variable into instance data structure `udc_ambiq_data` 

Thank you so much for pointing this out @jfischer-no